### PR TITLE
Alerting: State manager to use InstanceStore

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
Grafana database contains a table `alert_instance`. This table is used by alerting engine to persist the state of alerts between restarts. The table is exposed via InstanceStore repository. There are a few places where that repository is accessed. 
1. When Grafana starts it loads the information from the table to the state manager before evaluating the rules. This is the state manager's responsibility https://github.com/grafana/grafana/blob/085258c0351cb77126a87eae2421597145535f8e/pkg/services/ngalert/schedule/schedule.go#L283-L292 
2. When a rule is evaluated, the scheduler calls the state manager to process the evaluation results, and then the scheduler calls the InstanceStore to save the result provided by the state manager. https://github.com/grafana/grafana/blob/085258c0351cb77126a87eae2421597145535f8e/pkg/services/ngalert/schedule/schedule.go#L332
3.  When the scheduler is stopped (when Grafana stops, or an error occurs in the error group the scheduler is part of) it does many interactions with the database and state manager to flush the manager's state to the store. https://github.com/grafana/grafana/blob/085258c0351cb77126a87eae2421597145535f8e/pkg/services/ngalert/schedule/schedule.go#L283-L292

Therefore the InstanceStore ownership is divided between the scheduler and state manager but the state manager is the user and the source of the information provided by the repository. Therefore, it does not make sense to have a scheduler as a proxy. Moreover, the scheduler has to obtain information about the organizations and the request states for each in order to save them to the store.

This PR makes the state manager the only owner of the data stored provided by InstanceStore. This makes the boundaries clear, reduces the number of dependencies in the scheduler, and provides a way for further improvements. 

**Special notes for your reviewer:**
Also, I added some functions that would generate eval.Result. It will reduce the size of the tests

### CODE REVIEW

File: pkg/services/ngalert/state/manager.go [hunk 1]
```diff
@@ -297,6 +307,42 @@ func (st *Manager) Put(states []*State) {
...
+	t := st.clock.Now()
+	st.log.Info("flushing the state")
+	st.cache.mtxStates.Lock()
+	defer st.cache.mtxStates.Unlock()
+	totalStates, errorsCnt := 0, 0
+	for _, orgStates := range st.cache.states {
+		for _, ruleStates := range orgStates {
+			for _, state := range ruleStates {
+				err := st.saveState(ctx, state)
+				totalStates++
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/state/manager.go:321]: Should we increment `totalStates++` even if `err` is `non-nil`?

File: pkg/services/ngalert/state/manager.go [hunk 2]
```diff
@@ -297,6 +307,42 @@ func (st *Manager) Put(states []*State) {
...
+				if err != nil {
+					st.log.Error("failed to save alert state", append(state.GetRuleKey().LogContext(), "labels", state.Labels.String(), "state", state.State.String(), "err", err.Error()))
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/state/manager.go:323]: What does it mean to do a partial flush of the state? Should we not return err here instead?

**Review by grobinson-grafana — APPROVED** [review]

File: pkg/services/ngalert/state/manager.go (hunk 2 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/state/manager.go:323]: Good question that I asked myself as well. I moved the code as is. I think the idea was to save as many states as possible. The error here can come from instance validation or SQL.  I think skipping the individual state was motivated by the former error source.

File: pkg/services/ngalert/state/manager.go (hunk 1 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/state/manager.go:321]: I think so. The total states should reflect the total number of states. Together with the error counter, it provides the information.

**Pull request merged**